src/parser/src/SymbolTable.cpp

```diff
@@ -0,0 +1,142 @@
+                                     const std::string& funcNamespace) {
+    if (!this->data.contains(funcName)) {
+        throw std::runtime_error("[Internal Error] Could not update " + symbolName + " to new value function [" +
+                                 funcName + "] not found!");
+    }
+
+    std::string fullSymbolName = funcNamespace + "::" + symbolName;
+    if (!this->data[funcName].contains(fullSymbolName)) {
+        throw std::runtime_error("[Internal Error] Could not update " + symbolName + " to new value [" +
+                                 fullSymbolName + "] not found!");
+    }
+    this->data[funcName][fullSymbolName] = {this->data[funcName][fullSymbolName].type, value};
+}
+
+/**
+ * Returns true if the function is defined within the namespace
+ * @param funcName
+ * @param funcNamespace
+ * @return
+ */
+bool Utils::SymbolTable::isFunctionDefined(const std::string& funcName, const std::string& funcNamespace) {
+    return this->funcTable.contains(funcNamespace + "::" + funcName);
+}
+
+/**
+ * Sets the value of the function to be pointing to the llvm::Function object
+ * @param funcName
+ * @param params
+ * @param func
+ */
+void Utils::SymbolTable::setFunctionData(const std::string& funcName, std::vector<std::shared_ptr<Typing::Type>> params,
+                                         llvm::Function* func, const std::string& funcNamespace) {
+    // Safety checks
+    if (!isFunctionDefined(funcName, funcNamespace)) {
+        throw std::runtime_error("Function [" + funcName + "] not defined!");
+    }
+    const std::string fullFuncName = funcNamespace + "::" + funcName;
+    if (!this->funcTable[fullFuncName].contains(params)) {
+        // TODO: Either done in typing or made to actually report type issues
+        throw std::runtime_error("Function [" + fullFuncName + "] expected different parameters");
+    }
+
+    this->funcTable[fullFuncName][params] = {func};
+}
+
+/**
+ * Returns true if the function with the specified parameters is defined
+ * @param funcName
+ * @param params
+ * @return
+ */
+bool Utils::SymbolTable::isFunctionDefinedParam(const std::string& funcName,
+                                                const std::vector<std::shared_ptr<Typing::Type>>& params,
+                                                const std::string& funcNamespace) {
+    const std::string fullFuncName = funcNamespace + "::" + funcName;
+    if (!isFunctionDefined(funcName)) {
+        throw std::runtime_error("Function [" + funcName + "] not defined!");
+    }
+
+    return this->funcTable[fullFuncName].contains(params);
+}
+
+/**
+ * Retrieves the function corresponding to name and type
+ * @param funcName
+ * @param params
+ * @return
+ */
+Utils::FunctionTableEntry Utils::SymbolTable::getFunction(const std::string& funcName,
+                                                          std::vector<std::shared_ptr<Typing::Type>> params,
```

Comment:
This (and in addNewFunction, and setFunctionData) should probably a const ref to save copying a whole vector when its not needed

Reply:
Agreed, also forgot in isFunctionDefinedParam